src/GCP.h

```diff
@@ -101,97 +104,206 @@ class NotEnoughData : public std::runtime_error {
+    static inline std::map<GenotypeConfidence, GenotypePercentile> create_confidence_to_percentile_map (const std::vector<GenotypeConfidence> &unsorted_genotype_confidences);
+    static inline GenotypePercentile iterator_to_percentile(const std::vector<GenotypeConfidence>::const_iterator &it,
+            const std::vector<GenotypeConfidence> &genotype_confidences);
+    static inline double linear_interpolation(double x1, double x2, double y1, double y2, double x);
 };
+////////////////////////////////////////////////////////////////////////////////
+////////////////////////////////////////////////////////////////////////////////
+////////////////////////////////////////////////////////////////////////////////
+
+
+
+
+/*
+ * YOU HAVE REACHED THE END OF THE INTERFACE.
+ * BELOW ARE IMPLEMENTATION DETAILS USERS OF THIS LIBRARY DOES NOT NEED TO LOOK.
+ */
+
+////////////////////////////////////////////////////////////////////////////////
+////////////////////////////////////////////////////////////////////////////////
+////////////////////////////////////////////////////////////////////////////////
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+////////////////////////////////////////////////////////////////////////////////
+// Implementation details
+////////////////////////////////////////////////////////////////////////////////
+
+
+/* ________Model________ */
+template<typename ModelData>
+Model<ModelData>::Model(uint32_t seed) : random_number_generator(seed) {}
+
+
+/* ________Simulator________ */
+template<typename ModelData, typename Genotyper>
+Simulator<ModelData, Genotyper>::Simulator(Model<ModelData> * model) : model(model) {}
+
+template<typename ModelData, typename Genotyper>
+std::vector<GenotypeConfidence> Simulator<ModelData, Genotyper>::simulate(uint32_t iterations) {
+  std::vector<GenotypeConfidence> genotype_confidences;
+  genotype_confidences.reserve(iterations);
+
+  for (uint32_t iterations_done = 0; iterations_done < iterations; ++iterations_done) {
+    ModelData model_data = model->produce_data();
+    Genotyper genotyper(model_data);
+    GenotypeConfidence genotype_confidence = genotyper.get_genotype_confidence();
+    genotype_confidences.push_back(genotype_confidence);
+  }
+
+  std::sort(genotype_confidences.begin(), genotype_confidences.end());
+  return genotype_confidences;
+}
+
+
+/* ________Percentiler________ */
+Percentiler::Percentiler(const std::vector<GenotypeConfidence> &unsorted_genotype_confidences) {
+  bool enough_data = unsorted_genotype_confidences.size() >= 2;
+  if (not enough_data) {
+    throw NotEnoughData("Please provide at least two simulated genotype confidences.");
+  }
+  this->confidence_to_percentile = create_confidence_to_percentile_map(unsorted_genotype_confidences);
+
+}
+
+GenotypePercentile Percentiler::get_confidence_percentile(GenotypeConfidence queried_confidence) const {
+  auto upper_bound = confidence_to_percentile.upper_bound(queried_confidence);
```

Comment:
Ive just realised this has to be a call to `lower_bound` not upper bound. The latter always gives the first entry greater than the query. So we never return a known value. Linear interpolation is always called, and still gives the right result. But this is a bug (i think i may have introduced it).

Reply:
Really nice spotting! I refactored this function and did not see that. Fixed this here: https://github.com/leoisl/GCP/pull/10/commits/82bb64858e82477cd53f79ef4279db9467fbd528 (with two other smaller changes/fixes).

After these changes, one of the huge tests started failing. It turns out that I thought Minos had a bug, and I changed one of the answers to what I thought was correct. I actually got it wrong: Minos output was correct. This bugfix also fixed this previous bug, and now Minos output and our output are exactly the same for 4 tests (2 small and 2 huge tests) (commit fixing the tests: https://github.com/leoisl/GCP/pull/10/commits/41028b8c1a2ca86d8860abd817074c5751c7aaaa ).

Reply:
great!